File: algorithms/2_medium/I/00334_increasing_triplet_subsequence/o1_space_o_of_n_time.cpp
class Solution {
public:
    static bool increasingTriplet(const vector<int>& nums) {
        int mid_val = numeric_limits<int>::max();
        int min_val = numeric_limits<int>::max();

        for (const auto curr_val : nums) {
            if (curr_val > mid_val) {
                return true;
            }

            if (curr_val < min_val) {
                min_val = curr_val;
            } else if (curr_val > min_val && curr_val < mid_val) {
                mid_val = curr_val;
            }
        }

        return false;
    }
};

File: algorithms/2_medium/I/00334_increasing_triplet_subsequence/o1_space_o_of_n_time.rs
impl Solution {
    pub fn increasing_triplet(nums: Vec<i32>) -> bool {
        let mut mid_val = i32::MAX;
        let mut min_val = i32::MAX;

        for curr_val in nums {
            if curr_val > mid_val {
                return true;
            }

            if curr_val < min_val {
                min_val = curr_val;
            } else if curr_val > min_val && curr_val < mid_val {
                mid_val = curr_val;
            }
        }

        false
    }
}
